Refactor crypto interface: encapsulate hmac operations

drbild opened the pull request:
The `xtt` handshake context functions are fairly convoluted due to support for multiple suite specs. For example,
 * The handshake context initialization functions contain large switch statements (over the suite spec), with significant repetition.
 * The mapping of suite spec to implementing functions is defined in the handshake context functions, even though that mapping is more general.
 * The various crypto algorithms supported by various suites are all exposed (via unions) in the context structs.
 * The unions are defined at each usage site. This is repetitious and makes it very difficult to add a new algorithm or add compile-time defines to disable support for one algorithm.

This series contains two main refactoring to improve this:
1) It introduces a `suite_ops` struct to provide polymorphic access to the crypto primitives.  This allows the mapping of suite spec to functions to be defined in one place, independent of the context.

     The struct can eventually include fields for key exchange, attestation, AEAD, signing (identity key), and HMAC operations. In this series, just key exchange, AEAD, and HMAC operations are introduced.

     Further, support to override the default implementation can be added.  For example, a library user could chose to override the default software-based implementation of LRSW attestation with a TPM-based implementation.

1) The repeated `union` definitions that allow static declaration of crypto data types at compile time (before the specific algorithm is known at runtime) are extracted to one file.

There is significant amount of refactoring required before the `xtt` code is ready for external review or adding support to use the TPM for long-term keys. Rather than try to do it all at once (which will never happen), this PR takes the first step.

Notice the reduced line count and complexity in `context.h` and `context.c`.  However, the changes to `messages.c`, `crypto_utils.c`, `message_utils.c`, key_derivation.c`, and `signatures.c` are straightforward substitutions.

drbild committed:
crypto: add polymorphic interface for HMAC operations

drbild committed:
context: encapsulate all HMAC ops behind suite spec

drbild committed:
crypto: add polymorphic interface for key exchange operations

drbild committed:
context: encapsulate all key exchange ops behind suite spec

drbild committed:
crypto: remove deprecated x25519 wrapper methods

drbild committed:
crypto: add polymorphic interface for AEAD operations

drbild committed:
context: encapsulate all AEAD ops behind suite spec

drbild committed:
crypto: remove deprecated hash wrapper methods

drbild committed:
crypto: remove deprecated AEAD wrapper methods

drbild committed:
context: remove suite spec switch statements

All variable suite spec algorithms are now encapsulated behind a suite
ops interface. Thus, all cases of the giant switch statements in the
handshake context init functions are identical and can be removed.

Eventually the currentl fixed algorithms (LRSW for attestation and
ECDSA for signing keys) should be encapsulated behind suite ops
interfaces as well. For now they are left in the handshake context
struct.

zanebeckwith committed:
crypto: don't use outlen output param in hash

To be consistent with the other crypto API functions, which don't give
the caller the output length (and instead provide that information in
the ops struct), do the same with the hash function.

zanebeckwith commented:
I really like this. Thank you for starting this off.

I've made a few suggestions, as commits that I've pushed to your branch.

drbild commented:
@zanebeckwith Thanks for the fixes!  I've squashed most of them into the prior commits.

I made two primary changes:
1) `outlen` param to hash functions: keep as an input, to enable bounds checking when writing the hash output (since the hash functions take a plain `char *` for out, not a `struct xtt_crypto_hash...`). 
 This is more consistent with the `prf` signature, too.
2) Removed "desired" from the docstring for `outlen` arg in the `prf` function.  It is the size of the  `out` buffer (i.e., the max bytes to write).  Of course, that is implicitly also the desired bytes.

zanebeckwith committed:
Make hash function take struct, not raw pointer.
